crates/vm/src/expr.rs

```diff
@@ -1445,6 +1460,127 @@ impl QueryExpr {
         }
     }
 
+    /// Try to turn an inner join followed by a projection into a semijoin.
+    ///
+    /// This optimization recognizes queries of the form:
+    ///
+    /// ```ignore
+    /// QueryExpr {
+    ///   source: LHS,
+    ///   query: [
+    ///     JoinInner(JoinExpr {
+    ///       rhs: RHS,
+    ///       semi: false,
+    ///       ..
+    ///     }),
+    ///     Project(LHS.*),
+    ///     ...
+    ///   ]
+    /// }
+    /// ```
+    ///
+    /// And combines the `JoinInner` with the `Project` into a `JoinInner` with `semi: true`.
+    ///
+    /// Current limitations of this optimization:
+    /// - The `JoinInner` must be the first (0th) element of the `query`.
+    ///   Future work could search through the `query` to find any applicable `JoinInner`s,
+    ///   but the current implementation inspects only the first expr.
+    ///   This is likely sufficient because this optimization is primarily useful for enabling `try_index_join`,
+    ///   which is fundamentally limited to operate on the first expr.
+    /// - The `Project` must immediately follow the `JoinInner`, with no intervening exprs.
+    ///   Future work could search through intervening exprs to detect that the RHS table is unused.
+    /// - The LHS/source table must be a [`DbTable`], not a [`MemTable`].
```

Comment:
Is this what we want? We would like to apply this optimization to delta table queries.

Reply:
Expanded comment here.